### PR TITLE
Add HostAlias in KlusterletDeployOption

### PR DESCRIPTION
**Pull request by promacanthus**
#165 

To support registration-agent and work-agent use domain to communicate with hub api server.

### CODE REVIEW

File: operator/v1/types.go [hunk 1]
```diff
@@ -133,6 +133,11 @@ type KlusterletDeployOption struct {
 	// Note: Do not modify the Mode field once it's applied.
 	// +optional
 	Mode InstallMode `json:"mode"`
+
```
Comment (qiujian16) [inline operator/v1/types.go]: KlusterletDeployOption is to define whether klusterlet can be run in hosted or default mode. I would think HubApiServerHostAlias should be put in the spec. it cannot be `hub_api_server_host_alias`, it needs to follow the camel format such as "hubAPIServerHostAliase"

**Review by qiujian16 — COMMENTED** [review]
thanks, I understand the problem to solve. But I do not think adding host aliase field can fully resolve the issue. If you have an addon that needs to connect to the same hub, you will meet the same problem and need to take care of the host setup on addon deployment.

I wonder if it is better to resolve it in system wide, e.g. setting up a dns record in dns server of kube so we do not need to set it per deployment. FYI @elgnay I think we see this issue before.

File: operator/v1/types.go (hunk 1 above)
Comment (promacanthus) [reply operator/v1/types.go]: I had a wrong understanding of `KlusterletDeployOption`. I will put `HubApiServerHostAlias` in the spec and format as `hubAPIServerHostAlias`.

**Comment by elgnay** [comment]
Agreed with @qiujian16. Besides the registration agent and work agent, klusterlet operator itself also communicates with the hub kube apiserver (https://github.com/open-cluster-management-io/registration-operator/blob/803bfb374805414222e2aff669b73588dffe3537/pkg/operators/klusterlet/controllers/ssarcontroller/klusterlet_ssar_controller.go#L159),  it's a little bit complicated for klusterlet operator to handle the HostAlias in the klusterlet crs.

File: operator/v1/types.go [hunk 2]
```diff
@@ -327,6 +327,11 @@ type KlusterletSpec struct {
 	// RegistrationConfiguration contains the configuration of registration
 	// +optional
 	RegistrationConfiguration *RegistrationConfiguration `json:"registrationConfiguration,omitempty"`
+
+	// HubApiServerHostAlias contains the host alias for hub api server.
+	// registration-agent and work-agent will use it to communicate with hub api server.
+	// +optional
+	HubApiServerHostAlias *v1.HostAlias `json:"hubApiServerHostAlias,omitempty"`
```
Comment (qiujian16) [inline operator/v1/types.go]: does it need to be a pointer? what if user set one with empty IP or host name in it? we probably should not use v1 struct but have our own with more validation

Comment (promacanthus) [reply operator/v1/types.go]: I think a nil pointer indicates that agents should not use host alias. Add a new struct HubApiServerHostAlias with ip and hostname field. I think only need to set one ip and hostname for hub api server.

```golang
type HubApiServerHostAlias struct {
	// IP address of the host file entry.
	// +required
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Pattern=`^((0|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.){3}(0|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])$`
	IP string `json:"ip,omitempty" protobuf:"bytes,1,opt,name=ip"`

	// Hostname for the above IP address.
	// +required
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Pattern=`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`
	Hostname string `json:"hostnames,omitempty" protobuf:"bytes,2,rep,name=hostnames"`
}
```

File: operator/v1/types.go [hunk 3]
```diff
@@ -354,6 +359,22 @@ type NodePlacement struct {
...
 }
 
+// HubApiServerHostAlias holds the mapping between IP and hostnames that will be injected as an entry in the
+// pod's hosts file.
+type HubApiServerHostAlias struct {
+	// IP address of the host file entry.
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`^((0|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.){3}(0|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])$`
+	IP string `json:"ip,omitempty" protobuf:"bytes,1,opt,name=ip"`
```
Comment (qiujian16) [inline operator/v1/types.go]: no need to add protobuf, it is not supported in crd.

Comment (promacanthus) [reply operator/v1/types.go]: Done.   And i  think required field do not need omitempty tag.

**Comment by qiujian16** [comment]
thanks,I think it looks good. verify fails. Please also noted that klusterlet-operator itself will also use this host aliase to visit hub for status check. This is implementation details though.

**Comment by qiujian16** [comment]
btw, it will be good if you can add some tests https://github.com/open-cluster-management-io/api/blob/main/test/integration/klusterlet_test.go just to ensure the schema validation works as what your desired.

**Comment by promacanthus** [comment]
> thanks,I think it looks good. verify fails. Please also noted that klusterlet-operator itself will also use this host aliase to visit hub for status check. This is implementation details though.

Thanks. I will keep this in mind as i think about implementation details.

**Comment by promacanthus** [comment]
> btw, it will be good if you can add some tests https://github.com/open-cluster-management-io/api/blob/main/test/integration/klusterlet_test.go just to ensure the schema validation works as what your desired.

I will add some test cases to verify it.

**Comment by qiujian16** [comment]
@Promacanthus verify still fails. I think it should generate deepcopy but I do not see it in this PR.

**Comment by promacanthus** [comment]
> @Promacanthus verify still fails. I think it should generate deepcopy but I do not see it in this PR.

The problem has been found and fixed in the latest commit.

**Pull request merged**